File: clients/java/src/main/java/model/MineState.java
package model;

import util.StreamUtil;

public enum MineState {
    PREPARING(0),
    IDLE(1),
    TRIGGERED(2),
    EXPLODED(3);
    public int discriminant;
    MineState(int discriminant) {
      this.discriminant = discriminant;
    }
}

File: clients/scala/src/main/scala/model/MineState.scala
package model

import util.StreamUtil

sealed abstract class MineState (val discriminant: Int) {
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, discriminant)
    }
}

object MineState {
    case object PREPARING extends MineState(0)
    case object IDLE extends MineState(1)
    case object TRIGGERED extends MineState(2)
    case object EXPLODED extends MineState(3)
    def readFrom(stream: java.io.InputStream): MineState = StreamUtil.readInt(stream) match {
        case 0 => PREPARING
        case 1 => IDLE
        case 2 => TRIGGERED
        case 3 => EXPLODED
        case _ => throw new java.io.IOException("Unexpected discriminant value")
    }
}
